### PR TITLE
[33868] Refactor table events as output of the table component

### PR DESCRIPTION
**Pull request by oliverguenther**
Forward clicks on work packages and such to the parent components (WorkPackageListView and BCFListContainer), which can determine themselves what they want to do with it.

https://community.openproject.com/wp/33868

### CODE REVIEW

File: frontend/src/app/components/wp-table/timeline/container/wp-timeline-container.directive.ts [hunk 1]
```diff
@@ -60,6 +59,7 @@ import {HalEventsService} from "core-app/modules/hal/services/hal-events.service
 import {WorkPackageNotificationService} from "core-app/modules/work_packages/notifications/work-package-notification.service";
 import {combineLatest} from "rxjs";
 import {UntilDestroyedMixin} from "core-app/helpers/angular/until-destroyed.mixin";
+import {WorkPackagesTableComponent} from "core-components/wp-table/wp-table.component";
```
Comment (openprojectci) [inline frontend/src/app/components/wp-table/timeline/container/wp-timeline-container.directive.ts:63]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/0a4e473856cdedfb88250308f5ac62c11187065e/frontend/src/app/components/wp-table/timeline/container/wp-timeline-container.directive.ts#L63"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

**Comment by oliverguenther** [comment]
Hi @aleixsuau , I had some time and went ahead with the @Outputs refactoring for the table and card/grid view. In the latter case, events need to be forwarded twice (from WorkPackageCardView to WorkPackageGridView), but otherwise this appears to be working nicely for BIM/core functionality. I removed the overridden BCF Event Registry that handled clicking for showing viewpoints and moved that in the BCFListContainer.

Please have a look at the code to see where the changes ended up happening

File: frontend/src/app/components/wp-table/embedded/wp-embedded-table.component.ts [hunk 2]
```diff
@@ -12,6 +12,7 @@ import {QueryFormResource} from "core-app/modules/hal/resources/query-form-resou
 import {QueryFormDmService} from "core-app/modules/hal/dm-services/query-form-dm.service";
 import {distinctUntilChanged, map, take, withLatestFrom} from "rxjs/operators";
 import {InjectField} from "core-app/helpers/angular/inject-field.decorator";
+import {KeepTabService} from "core-components/wp-single-view-tabs/keep-tab/keep-tab.service";
```
Comment (openprojectci) [inline frontend/src/app/components/wp-table/embedded/wp-embedded-table.component.ts:16]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/0a4e473856cdedfb88250308f5ac62c11187065e/frontend/src/app/components/wp-table/embedded/wp-embedded-table.component.ts#L16"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

**Comment by oliverguenther** [comment]
As it turns out, this results in quite a bit of duplication whenever the behavior should be the same (i.e. double click => move to full view). This happens in board, wp view and embedded table components. That's probably what motivated me to use the registry that can be overridden. It's less discoverable. Let me know what you think Aleix

**Comment by aleixsuau** [comment]
Very nice refactor @oliverguenther , emitting actions from the table looks more declarative and simple to use. The only improvements that I could suggest are:

- Remove the 'on' prefix on the @Output names (https://angular.io/guide/styleguide#style-05-16). The general practice is to call the eventHandler with the 'on' prefix but not the @Output itself: <table (itemClicked)="OnItemClicked($event)"></table>
- Modify the UI to handle the cases where there are no actions (@Output bindings) when the need arises. For example when there is no @itemClicked event, then the link should be removed from the wp id.

Otherwise LGTM 👌 🙆‍♂️

**Comment by oliverguenther** [comment]
Thanks @aleixsuau , I addressed the first - we should maybe introduce the angular styleguide officially here as a reference, something that we did not do in the past.

The latter point is not going to be easy since the output component is very disconnected from the actual rendering of the table. As we discussed, I'd rather we ensure that outputs are checked on the component through some way to avoid using the table without these outputs being used, or introduce a default action instead. I'm not set on either, maybe we should do that on a subsequent PR then

**Pull request merged**